### PR TITLE
Add auth support for UAA IDP

### PR DESCRIPTION
**Pull request by vuil**
A fair bit of code refactoring to support multiple IDPs, include new support for UAA.

IDP used in CLI context creation is stored in the Context's additionalMetadata map.

`context get-token` is updated to refresh token based on the IDP associated with said context

`tanzu context create / tanzu login`
are updated to use UAA IDP unless the CLI recognizes the endpoint as a public Tanzu Platform service endpoint or if CSP idp is explicitly requested with the `--force-csp` hidden flag.

### Which issue(s) this PR fixes

Fixes #

### Describe testing done for PR

- Ran unit tests

- manual testings of following commands:

```
1.  > tanzu login --staging --endpoint https://xx--tanzu-hub-endpoint https://yyy

2  > tanzu login 

3. > tanzu login --endpoint https://local-uaa-based-endpoint

4. > ./bin/tanzu context create mytanzuUAA5 -t tanzu --endpoint https://www.alternate.endpoint.com
  [i] Opening the browser window to complete the login
  Log in by visiting this link:

      https://www.alternate.endpoint.com/auth/oauth/authorize?client_id=tp_cli_app&code_challenge=dKfHWRO5hsIVCDFUCDq1g9SjfasXPAeA1rzim7DSDt0&code_challenge_method=S256&redirect_uri=http%3A%2F%2F127.0.0.1%3A57080%2Fcallback&response_type=code&state=575286858256d936f902d73cfe53af31

      Optionally, paste your authorization code: [...]


  [ok] Successfully logged in to 'https://www.alternate.endpoint.com' and created a tanzu context
  > tanzu context current
    Name:            mytanzuUAA5
    Type:            tanzu
    Organization:     ()
    Project:         none set
    Kube Config:     /Users/johndoe/.config/tanzu/kube/config
    Kube Context:    tanzu-cli-mytanzuUAA5

  > ./bin/tanzu login --endpoint https://www.alternate.endpoint.com
  [i] Opening the browser window to complete the login
  Log in by visiting this link:

      https://www.alternate.endpoint.com/auth/oauth/authorize?client_id=tp_cli_app&code_challenge=Zk567ZHRL21Kkp0RpBRbz7QnGaSw9TlD7szg27sHAr4&code_challenge_method=S256&redirect_uri=http%3A%2F%2F127.0.0.1%3A57333%2Fcallback&response_type=code&state=74259704d8911cc27469cb10127f5b29

      Optionally, paste your authorization code: [...]


  [ok] Successfully logged in to 'https://www.alternate.endpoint.com' and created a tanzu context

  > tanzu context current
    Name:            tpsm-c297cb71
    Type:            tanzu
    Project:         none set
    Kube Config:     /Users/johndoe/.config/tanzu/kube/config
    Kube Context:    tanzu-cli-tpsm-c297cb71

5. > tanzu login --endpoint https://dev-csp-based-endpoint --force-csp

6. > tanzu context get-token context_name (for both CSP and UAA-based contexts)
```

### Release note

```release-note
Add support for using UAA as IDP. After this change, endpoint URIs provided when creating a 'tanzu' context not recognized as public Tanzu SaaS endpoint will default to use UAA as IDP. Use --force-csp boolean flag to override behavior for custom endpoints that still requires CSP.
```

### CODE REVIEW

File: pkg/command/context.go [hunk 1]
```diff
@@ -581,20 +606,73 @@ func createContextWithTanzuEndpoint() (context *configtypes.Context, err error)
 	}
 
 	sanitizedEndpoint := sanitizeEndpoint(endpoint)
+	idpType := getIDPType(sanitizedEndpoint)
+	endpoints := getEndpoints(sanitizedEndpoint, idpType)
+	fmt.Printf("Create context %s: idpType = %v, endpoints: %+v\n", ctxName, idpType, endpoints)
```
Comment (prkalle) [inline pkg/command/context.go:611]: nit: Can we removed this line be removed or is it intentional?

**Review by prkalle — COMMENTED** [review]
Nice refactoring @vuil ! 
I am still continuing the review, but sharing few comments i have so far. Thanks

File: pkg/command/context.go (hunk 1 above)
Comment (vuil) [reply pkg/command/context.go]: removed

File: pkg/auth/csp/tanzu.go [hunk 2]
```diff
@@ -618,14 +257,11 @@ func getCSPMetadataFromCentralConfig() TanzuCSPMetadata {
 // getCSPKnownIssuersEndpoints gets the CSP known issuer endpoints from central config as best effort,
 // If it fails to fetch data from central config, it returns the default values
 func getCSPKnownIssuersEndpoints() cspKnownIssuerEndpoints {
-	cspKnownIssuerEPs, err := getCSPKnownEndpointsFromCentralConfig()
-	if err == nil {
-		return cspKnownIssuerEPs
-	}
```
[the remark this answers is not in the document]
Comment (vuil) [reply pkg/auth/csp/tanzu.go:624]: good catch, this was left behind from testing the local issue a while ago. Undone.

**Review by prkalle — COMMENTED** [review]
Changes LGTM, just couple  more comments and then good to go.

File: pkg/command/context.go [hunk 3]
```diff
@@ -620,6 +638,38 @@ func globalLogin(c *configtypes.Context) (err error) {
...
+	claims, err := doInteractiveLoginAndUpdateContext(c, uaaEndpoint)
+	if err != nil {
+		return err
+	}
+
+	// UAA-based authentication does not provide org id or name yet
+	orgName := ""
+
+	if err := updateContextOnTanzuLogin(c, generateContextNameFunc, claims, orgName); err != nil {
+		return nil
```
[the remark this answers is not in the document]
Comment (vuil) [reply pkg/command/context.go]: oops, thanks.

File: pkg/command/context.go [hunk 4]
```diff
@@ -620,6 +638,38 @@ func globalLogin(c *configtypes.Context) (err error) {
...
+	uaaEndpoint := c.AdditionalMetadata[config.TanzuAuthEndpointKey].(string)
+	log.V(7).Infof("Login to UAA endpoint: %s", uaaEndpoint)
+
+	claims, err := doInteractiveLoginAndUpdateContext(c, uaaEndpoint)
+	if err != nil {
+		return err
+	}
+
+	// UAA-based authentication does not provide org id or name yet
+	orgName := ""
```
Comment (anujc25) [inline pkg/command/context.go:659]: OrgName which gets configured in the `context` as part of additionalMetadata is used by all the plugins to show log messages. Keeping this empty might make the log message incomplete.

Comment (anujc25) [reply pkg/command/context.go:659]: We can do this as part of follow up as well.

**Review by anujc25 — APPROVED** [review]

**Review by prkalle — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**